extensions/amp-a4a/0.1/name-frame-renderer.js

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Renderer} from './amp-ad-type-defs';
+import {createElementWithAttributes} from '../../../src/dom';
+import {dev} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {getContextMetadata} from '../../../src/iframe-attributes';
+import {getDefaultBootstrapBaseUrl} from '../../../src/3p-frame';
+import {utf8Decode} from '../../../src/utils/bytes';
+
+/**
+ * Render a non-AMP creative into a NameFrame.
+ */
+export class NameFrameRenderer extends Renderer {
+  /** @override */
+  render(context, element, unusedCreativeData) {
+    const crossDomainData = context.crossDomainData;
+    dev().assert(crossDomainData, 'CrossDomain data undefined!');
+
+    const creative = utf8Decode(crossDomainData.rawCreativeBytes);
+    const srcPath =
+        getDefaultBootstrapBaseUrl(context.win, 'nameframe');
+    const contextMetadata = getContextMetadata(
+        context.win,
+        element,
+        crossDomainData.sentinel,
+        crossDomainData.additionalContextMetadata);
```

Comment:
where are those `additionalContextMetadata` generated?

Reply:
By the subclasses of AmpAdNetworkBase that wish to use the NameFrameRenderer.

Reply:
do we have example of those attributes? Want to see if those are ad network specific, or can be shared in NameFrameRenderer.

Reply:
They are ad network specific. For example, fluid ads have a different set of attributes than other non-fluid DoubleClick creatives rendered in safeframe.